src/osemgrep/reporting/Nosem.ml

```diff
@@ -69,11 +70,96 @@ let _nosem_previous_line_re =
 (* Entry point *)
 (*****************************************************************************)
 
-(* TODO: probably need to pass some flags from the CLI *)
-let process_ignores (out : Out.cli_output) : Out.cli_output =
-  (* TODO: port nosemgrep.py, filter in out.matches the
-     matches where the match has a corresponding nosem: annotation in
-     the corresponding file, and add new errors if nosem: has the wrong
-     format in the file.
-  *)
-  out
+let rule_match_nosem ~strict (rule_match : Out.cli_match) :
+    bool * Out.cli_error list =
+  let lines =
+    File.lines_of_file
+      (max 0 (rule_match.Out.start.line - 1), rule_match.Out.end_.line)
+      rule_match.Out.path
+  in
+
+  let previous_line, line =
+    match lines with
+    | line0 :: line1 :: _ when rule_match.Out.start.line > 0 ->
+        (Some line0, Some line1)
+    | line :: _ -> (None, Some line)
+    | [] (* XXX(dinosaure): is it possible? *) -> (None, None)
+  in
+
+  let recognise_and_collect ~rex line =
```

Comment:
avoid nested functions. Move out in Helper section, and give type and small explanation if needed.